### PR TITLE
Modification of reward signals and rl_trainer for SAC

### PR DESCRIPTION
**Pull request by ervteng**
Had to make additional changes to reward signals and rl_trainer to abstract to off-policy.
- Adds `evaluate_batch` to reward signals. Evaluates on minibatch rather than on BrainInfo.
- Changes the way reward signal results are reported in rl_trainer so that we get the pure, unprocessed environment reward separate from the reward signals. 
- Moves `end_episode` to rl_trainer

### CODE REVIEW

File: ml-agents/mlagents/trainers/rl_trainer.py [hunk 1]
```diff
@@ -10,10 +10,23 @@
...
 
 LOGGER = logging.getLogger("mlagents.trainers")
 
+RewardSignalResults = Dict[str, RewardSignalResult]
+
+
+class AllRewardsOutput(NamedTuple):
+    """
+    CLass stores outputs of reward signals on current BrainInfos.
+    Also stores the actual rewards and optionally the values.
```
[the remark this answers is not in the document]
Comment (ervteng) [reply ml-agents/mlagents/trainers/rl_trainer.py]: Originally had it then realized it didn't make sense. Edited the comment for this.

File: ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py [hunk 2]
```diff
@@ -62,31 +62,51 @@ def __init__(
...
         self, current_info: BrainInfo, next_info: BrainInfo
     ) -> RewardSignalResult:
         if len(current_info.agents) == 0:
-            return []
+            return RewardSignalResult([], [])
+        mini_batch: Dict[str, np.array] = {}
+        # Construct the batch
+        mini_batch["actions"] = next_info.previous_vector_actions
+        mini_batch["done"] = np.reshape(next_info.local_done, [-1, 1])
+        for i, _ in enumerate(current_info.visual_observations):
```
[the remark this answers is not in the document]
Comment (ervteng) [reply ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py]: Changed the enumerates to either `range(len(` or using the actual second term if it's used.

**Comment by harperj** [comment]
Latest changes LGTM.  Can you briefly summarize how you tested this PR?

**Comment by ervteng** [comment]
> Latest changes LGTM. Can you briefly summarize how you tested this PR?

Tested GAIL with Crawler and Curiosity + GAIL with Pyramids. Tested GAIL with GridWorld for visual obs.

**Pull request merged**